### PR TITLE
chore: Use `tracing` instead of `log` and `simplelog`

### LINKED ISSUE #177: Use tracing crate for logging
We now use env_logger. In the DI sanitizer we do some tree walking and have a pretty bad hierarchical log thing. We should switch to tracing so we get proper structured, hierarchical logging.

### PR DESCRIPTION
**Pull request by vadorovsky**
Structured logging from `tracing` meets our needs better, especially in the parts of bpf-linker which sanitize the debug info.

Fixes #177

- - -
This change is [Reviewable](https://reviewable.io/reviews/aya-rs/bpf-linker/179)

### CODE REVIEW

**Review by tamird — CHANGES_REQUESTED** [review]
> Reviewed 7 of 7 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/aya-rs/bpf-linker/179)* status: all files reviewed, 9 unresolved discussions (waiting on @alessandrod and @vadorovsky)

___
*[`Cargo.toml` line 31 at r1](https://reviewable.io/reviews/aya-rs/bpf-linker/179#-NpoW_KECvhWDFQJYqO6:-NpoW_KECvhWDFQJYqO7:bbezqem) ([raw file](https://github.com/aya-rs/bpf-linker/blob/1f6476981267a1992d4a718b0c2a94b256ee7624/Cargo.toml#L31)):*
> ```TOML
> tracing = "0.1"
> tracing-appender = "0.2"
> tracing-subscriber = { version = "0.3", features = ["env-filter", "json", "registry"] }
> ```

Should this be above, or is it used by lib?
___
*[`src/bin/bpf-linker.rs` line 16 at r1](https://reviewable.io/reviews/aya-rs/bpf-linker/179#-NpoXtM81EqSLHFRPENy:-NpoXtM81EqSLHFRPENz:bxbdjm1) ([raw file](https://github.com/aya-rs/bpf-linker/blob/1f6476981267a1992d4a718b0c2a94b256ee7624/src/bin/bpf-linker.rs#L16)):*
> ```XML
> use tracing_tree::HierarchicalLayer;
> 
> const TRACING_IDENT: usize = 2;
> ```

move this into the function where it's used, or just inline "2" there?
___
*[`src/bin/bpf-linker.rs` line 65 at r1](https://reviewable.io/reviews/aya-rs/bpf-linker/179#-NpoWlJt4hr5LLSYSuZU:-NpoWlJt4hr5LLSYSuZV:b-1d86zm) ([raw file](https://github.com/aya-rs/bpf-linker/blob/1f6476981267a1992d4a718b0c2a94b256ee7624/src/bin/bpf-linker.rs#L65)):*
> ```XML
> 
> #[derive(Clone, Debug)]
> enum LogLevel {
> ```

Do we need this? Can we use tracing::Level? That type has a `FromStr` impl.
___
*[`src/bin/bpf-linker.rs` line 258 at r1](https://reviewable.io/reviews/aya-rs/bpf-linker/179#-NpoYGyV162KQ849JZZC:-NpoYGyV162KQ849JZZD:bwqca2v) ([raw file](https://github.com/aya-rs/bpf-linker/blob/1f6476981267a1992d4a718b0c2a94b256ee7624/src/bin/bpf-linker.rs#L258)):*
> ```XML
>     ) = match log_file {
>         Some(log_file) => {
>             let file_appender = tracing_appender::rolling::daily(
> ```

do we need this level of sophistication?
___
*[`src/llvm/di.rs` line 213 at r1](https://reviewable.io/reviews/aya-rs/bpf-linker/179#-NpoYblC2dJFwvkcrEiS:-NpoYblC2dJFwvkcrEiT:b5kve1e) ([raw file](https://github.com/aya-rs/bpf-linker/blob/1f6476981267a1992d4a718b0c2a94b256ee7624/src/llvm/di.rs#L213)):*
> ```XML
>         let item_span = span!(Level::TRACE, "item", value_id);
>         let _enter = item_span.enter();
>         trace!(item = ?item, value = ?value_ref, "visiting item");
> ```

i think `item = ?item` is the same as `?item`?
___
*[`src/llvm/di.rs` line 80 at r2](https://reviewable.io/reviews/aya-rs/bpf-linker/179#-NpoYxqD9cpUXX53emhV:-NpoYxqD9cpUXX53emhW:b-jt5cni) ([raw file](https://github.com/aya-rs/bpf-linker/blob/44b1e400534cdf9eafbb64a9342cb991b37f7d8d/src/llvm/di.rs#L80)):*
> ```XML
>                 match di_composite_type.tag() {
>                     DW_TAG_structure_type => {
>                         let (original_name, sanitized_name) = match di_composite_type.name() {
> ```

is there a reason these both need to options rather than a single option of a tuple? seems they're either both some or both none.
___
*[`src/llvm/di.rs` line 82 at r2](https://reviewable.io/reviews/aya-rs/bpf-linker/179#-NpoYpxJ8IDuK3gt-bLC:-NpoYpxJ8IDuK3gt-bLD:bpa07gf) ([raw file](https://github.com/aya-rs/bpf-linker/blob/44b1e400534cdf9eafbb64a9342cb991b37f7d8d/src/llvm/di.rs#L82)):*
> ```XML
>                         let (original_name, sanitized_name) = match di_composite_type.name() {
>                             Some(name) => {
>                                 let original_name = name.to_string_lossy().to_string();
> ```

can we avoid the to_string call?
___
*[`src/llvm/di.rs` line 177 at r2](https://reviewable.io/reviews/aya-rs/bpf-linker/179#-NpoZCNx5dxpzxJBF_Xk:-NpoZCNx5dxpzxJBF_Xl:b-t9uma7) ([raw file](https://github.com/aya-rs/bpf-linker/blob/44b1e400534cdf9eafbb64a9342cb991b37f7d8d/src/llvm/di.rs#L177)):*
> ```XML
>                         }
>                         if remove_name {
>                             // If the type was marked with `AyaBtfMapMarker`,
> ```

weird comment given we don't know anything about `AyaBtfMapMarker` here.
___
*[`tests/assembly/di_generics.rs` line 2 at r1](https://reviewable.io/reviews/aya-rs/bpf-linker/179#-NpoYjCJ6kqVZNS1sMOz:-NpoYjCJ6kqVZNS1sMP-:bjpuaix) ([raw file](https://github.com/aya-rs/bpf-linker/blob/1f6476981267a1992d4a718b0c2a94b256ee7624/tests/assembly/di_generics.rs#L2)):*
> ```XML
> // assembly-output: bpf-linker
> // compile-flags: --crate-type cdylib -C link-arg=--emit=llvm-ir -C link-arg=--btf -C debuginfo=2 -C link-arg=--log-file=/tmp/bpf-linker
> ```

is it appropriate to log to a file here?

**Review by vadorovsky — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/aya-rs/bpf-linker/179)* status: all files reviewed, 9 unresolved discussions (waiting on @alessandrod and @tamird)

___
*[`Cargo.toml` line 31 at r1](https://reviewable.io/reviews/aya-rs/bpf-linker/179#-NpoW_KECvhWDFQJYqO6:-Npo_8RRBIisj5c32iFX:b8dhjxx) ([raw file](https://github.com/aya-rs/bpf-linker/blob/1f6476981267a1992d4a718b0c2a94b256ee7624/Cargo.toml#L31)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Should this be above, or is it used by lib?
</blockquote></details>

Yes, sorry, all the new libraries (except just `tracing`) are used only in bin.
___
*[`src/bin/bpf-linker.rs` line 16 at r1](https://reviewable.io/reviews/aya-rs/bpf-linker/179#-NpoXtM81EqSLHFRPENy:-NpogNIwE6zA4baW5qv-:b-896fix) ([raw file](https://github.com/aya-rs/bpf-linker/blob/1f6476981267a1992d4a718b0c2a94b256ee7624/src/bin/bpf-linker.rs#L16)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

move this into the function where it's used, or just inline "2" there?
</blockquote></details>

Done.
___
*[`src/bin/bpf-linker.rs` line 65 at r1](https://reviewable.io/reviews/aya-rs/bpf-linker/179#-NpoWlJt4hr5LLSYSuZU:-Npo_HEg8mzpgTamR7kb:bxdlo8x) ([raw file](https://github.com/aya-rs/bpf-linker/blob/1f6476981267a1992d4a718b0c2a94b256ee7624/src/bin/bpf-linker.rs#L65)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Do we need this? Can we use tracing::Level? That type has a `FromStr` impl.
</blockquote></details>

Not really - tracing::Level is not an enum, it's a struct https://docs.rs/tracing/0.1.40/tracing/struct.Level.html. So I can't use it directly with clap. This enum serves as a translation from clap argument to tracing::Level.
___
*[`src/bin/bpf-linker.rs` line 258 at r1](https://reviewable.io/reviews/aya-rs/bpf-linker/179#-NpoYGyV162KQ849JZZC:-Npob2T39-iVp1iJMv4g:b-oq0jlf) ([raw file](https://github.com/aya-rs/bpf-linker/blob/1f6476981267a1992d4a718b0c2a94b256ee7624/src/bin/bpf-linker.rs#L258)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

do we need this level of sophistication?
</blockquote></details>

Changed to `rolling::never`
___
*[`src/llvm/di.rs` line 213 at r1](https://reviewable.io/reviews/aya-rs/bpf-linker/179#-NpoYblC2dJFwvkcrEiS:-NpogJMv2o-lnhqqX2kc:b-896fix) ([raw file](https://github.com/aya-rs/bpf-linker/blob/1f6476981267a1992d4a718b0c2a94b256ee7624/src/llvm/di.rs#L213)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

i think `item = ?item` is the same as `?item`?
</blockquote></details>

Done.
___
*[`src/llvm/di.rs` line 80 at r2](https://reviewable.io/reviews/aya-rs/bpf-linker/179#-NpoYxqD9cpUXX53emhV:-Npob9v1EARnU-h8oNjv:b-gdn3k8) ([raw file](https://github.com/aya-rs/bpf-linker/blob/44b1e400534cdf9eafbb64a9342cb991b37f7d8d/src/llvm/di.rs#L80)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

is there a reason these both need to options rather than a single option of a tuple? seems they're either both some or both none.
</blockquote></details>

They are used separately in different places (the C-compatible name is being written to DI, the Rust name is just being logged). If I did as you say, instead of doing

```rust
if let Some(sanitized_name) = sanitized_name {
    di_composite_type
        .replace_name(self.context, sanitized_name.as_str())
        .unwrap();
}
```

I would need to do

```rust
if let Some(name_tuple) = name_tuple {
    di_composite_type
        .replace_name(self.context, name_tuple.0)
        .unwrap();
```

I don't really like the second example.
___
*[`src/llvm/di.rs` line 82 at r2](https://reviewable.io/reviews/aya-rs/bpf-linker/179#-NpoYpxJ8IDuK3gt-bLC:-Npo_of3D6e2F1PSPCW2:bvaiofj) ([raw file](https://github.com/aya-rs/bpf-linker/blob/44b1e400534cdf9eafbb64a9342cb991b37f7d8d/src/llvm/di.rs#L82)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

can we avoid the to_string call?
</blockquote></details>

I would say, not really.

Keeping it as `Cow<'_, str>`(what `.to_string_lossy()` returns) has quite annoying consequences:

* I would need to change the type of `skipped_types` field to `Vec<Cow<'a, str>>` (which would also introduce a lifetime to the whole struct) or `Cow<'static, str>`(well, not sure if that's even possible).
* Either way, this triggers nasty borrowing issues.

If I do:

```diff
diff --git a/src/llvm/di.rs b/src/llvm/di.rs
index 35ce1df..9a631d0 100644
--- a/src/llvm/di.rs
+++ b/src/llvm/di.rs
@@ -21,14 +21,14 @@ use crate::llvm::{iter::*, types::di::DISubprogram};
 // backward compatibility
 const MAX_KSYM_NAME_LEN: usize = 128;

-pub struct DISanitizer {
+pub struct DISanitizer<'a> {
     context: LLVMContextRef,
     module: LLVMModuleRef,
     builder: LLVMDIBuilderRef,
     visited_nodes: HashSet<u64>,
     item_stack: Vec<Item>,
     replace_operands: HashMap<u64, LLVMMetadataRef>,
-    skipped_types: Vec<String>,
+    skipped_types: Vec<Cow<'a, str>>,
 }

 // Sanitize Rust type names to be valid C type names.
@@ -59,8 +59,8 @@ fn sanitize_type_name<T: AsRef<str>>(name: T) -> String {
     n
 }

-impl DISanitizer {
-    pub fn new(context: LLVMContextRef, module: LLVMModuleRef) -> DISanitizer {
+impl<'a> DISanitizer<'a> {
+    pub fn new(context: LLVMContextRef, module: LLVMModuleRef) -> DISanitizer<'a> {
         DISanitizer {
             context,
             module,
@@ -81,7 +81,7 @@ impl DISanitizer {
                     DW_TAG_structure_type => {
                         let (original_name, sanitized_name) = match di_composite_type.name() {
                             Some(name) => {
-                                let original_name = name.to_string_lossy().to_string();
+                                let original_name = name.to_string_lossy();
                                 let sanitized_name = sanitize_type_name(&original_name);

                                 (Some(original_name), Some(sanitized_name))
@@ -115,7 +115,7 @@ impl DISanitizer {

                                             let name = match original_name {
                                                 Some(name) => name,
-                                                None => "(anon)".to_owned(),
+                                                None => Cow::Borrowed("(anon)"),
                                             };
                                             let filename = match filename {
                                                 Some(filename) => {
```

One of the annoying errors that occur is:

```
error[E0502]: cannot borrow `di_composite_type` as mutable because it is also borrowed as immutable
   --> src/llvm/di.rs:176:29
    |
62  | impl<'a> DISanitizer<'a> {
    |      -- lifetime `'a` defined here
...
82  |                         let (original_name, sanitized_name) = match di_composite_type.name() {
    |                                                                     ----------------- immutable borrow occurs here
...
130 |                                             self.skipped_types.push(name);
    |                                             ----------------------------- argument requires that `di_composite_type` is borrowed for `'a`
...
176 |                             di_composite_type.replace_elements(sorted_elements);
    |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ mutable borrow occurs here
```

Well, the obvious solution would be pushing owned Strings to the vector to not cause an immutable borrow. Which pretty much brings us back to what I'm doing.

And we are not even benefiting from copy-on-write in this code anyway.
___
*[`src/llvm/di.rs` line 177 at r2](https://reviewable.io/reviews/aya-rs/bpf-linker/179#-NpoZCNx5dxpzxJBF_Xk:-Npoa17U8EA1fiibdTZq:b-4gfgb6) ([raw file](https://github.com/aya-rs/bpf-linker/blob/44b1e400534cdf9eafbb64a9342cb991b37f7d8d/src/llvm/di.rs#L177)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

weird comment given we don't know anything about `AyaBtfMapMarker` here.
</blockquote></details>

It's explained in an another comment above (line 148). Do you think it's worth repeating here?
___
*[`tests/assembly/di_generics.rs` line 2 at r1](https://reviewable.io/reviews/aya-rs/bpf-linker/179#-NpoYjCJ6kqVZNS1sMOz:-NpoaJ4m2rPq6YFmGN1l:b-hx8wl3) ([raw file](https://github.com/aya-rs/bpf-linker/blob/1f6476981267a1992d4a718b0c2a94b256ee7624/tests/assembly/di_generics.rs#L2)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

is it appropriate to log to a file here?
</blockquote></details>

Whoops, nope, removed

File: src/bin/bpf-linker.rs [hunk 1]
```diff
@@ -110,8 +152,8 @@ struct CommandLine {
     log_file: Option<PathBuf>,
 
     /// Set the log level. Can be one of `off`, `info`, `warn`, `debug`, `trace`.
```
Comment (dave-tucker) [inline src/bin/bpf-linker.rs:154]: Is this doc comment still accurate? I don't see `off` in the enum + `error` is missing from the doc comment.

File: src/bin/bpf-linker.rs [hunk 2]
```diff
@@ -191,51 +245,37 @@ fn main() {
...
-    let env_log_level = match env::var("RUST_LOG") {
-        Ok(s) if !s.is_empty() => match s.parse::<LevelFilter>() {
-            Ok(l) => Some(l),
-            Err(e) => error(
-                &format!("invalid RUST_LOG value: {e}"),
-                clap::error::ErrorKind::InvalidValue,
-            ),
-        },
-        _ => None,
+    // Configure tracing.
```
Comment (dave-tucker) [inline src/bin/bpf-linker.rs:248]: I think this need plumbing for [`OFF`](https://docs.rs/tracing/latest/tracing/level_filters/struct.LevelFilter.html#associatedconstant.OFF) so we can disable the tracing entirely.

**Review by dave-tucker — CHANGES_REQUESTED** [review]
Couple of nits but overall this change is looking good! Thanks.

**Review by tamird — CHANGES_REQUESTED** [review]
> Reviewed 5 of 5 files at r5, 2 of 2 files at r6, 2 of 2 files at r7, 2 of 2 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/aya-rs/bpf-linker/179)* status: all files reviewed, 6 unresolved discussions (waiting on @alessandrod and @vadorovsky)

___
*[`src/bin/bpf-linker.rs` line 65 at r1](https://reviewable.io/reviews/aya-rs/bpf-linker/179#-NpoWlJt4hr5LLSYSuZU:-Npou5ocEmhr2IDloGzd:b-5g2xwf) ([raw file](https://github.com/aya-rs/bpf-linker/blob/1f6476981267a1992d4a718b0c2a94b256ee7624/src/bin/bpf-linker.rs#L65)):*
<details><summary><i>Previously, vadorovsky (Michal Rostecki) wrote…</i></summary><blockquote>

Not really - tracing::Level is not an enum, it's a struct https://docs.rs/tracing/0.1.40/tracing/struct.Level.html. So I can't use it directly with clap. This enum serves as a translation from clap argument to tracing::Level.
</blockquote></details>

Why can't you use it directly? Isn't `FromStr` all you need?
___
*[`src/llvm/di.rs` line 80 at r2](https://reviewable.io/reviews/aya-rs/bpf-linker/179#-NpoYxqD9cpUXX53emhV:-NpouTrx2QqdqgrpySYb:b-v0if4s) ([raw file](https://github.com/aya-rs/bpf-linker/blob/44b1e400534cdf9eafbb64a9342cb991b37f7d8d/src/llvm/di.rs#L80)):*
<details><summary><i>Previously, vadorovsky (Michal Rostecki) wrote…</i></summary><blockquote>

They are used separately in different places (the C-compatible name is being written to DI, the Rust name is just being logged). If I did as you say, instead of doing

```rust
if let Some(sanitized_name) = sanitized_name {
    di_composite_type
        .replace_name(self.context, sanitized_name.as_str())
        .unwrap();
}
```

I would need to do

```rust
if let Some(name_tuple) = name_tuple {
    di_composite_type
        .replace_name(self.context, name_tuple.0)
        .unwrap();
```

I don't really like the second example.
</blockquote></details>

You could write `if let Some((_, foo)) = ...` which seems fine
___
*[`src/llvm/di.rs` line 82 at r2](https://reviewable.io/reviews/aya-rs/bpf-linker/179#-NpoYpxJ8IDuK3gt-bLC:-NpouksJ-sEwjX95Ci2V:b-4q725j) ([raw file](https://github.com/aya-rs/bpf-linker/blob/44b1e400534cdf9eafbb64a9342cb991b37f7d8d/src/llvm/di.rs#L82)):*
<details><summary><i>Previously, vadorovsky (Michal Rostecki) wrote…</i></summary><blockquote>

I would say, not really.

Keeping it as `Cow<'_, str>`(what `.to_string_lossy()` returns) has quite annoying consequences:

* I would need to change the type of `skipped_types` field to `Vec<Cow<'a, str>>` (which would also introduce a lifetime to the whole struct) or `Cow<'static, str>`(well, not sure if that's even possible).
* Either way, this triggers nasty borrowing issues.

If I do:

```diff
diff --git a/src/llvm/di.rs b/src/llvm/di.rs
index 35ce1df..9a631d0 100644
--- a/src/llvm/di.rs
+++ b/src/llvm/di.rs
@@ -21,14 +21,14 @@ use crate::llvm::{iter::*, types::di::DISubprogram};
 // backward compatibility
 const MAX_KSYM_NAME_LEN: usize = 128;

-pub struct DISanitizer {
+pub struct DISanitizer<'a> {
     context: LLVMContextRef,
     module: LLVMModuleRef,
     builder: LLVMDIBuilderRef,
     visited_nodes: HashSet<u64>,
     item_stack: Vec<Item>,
     replace_operands: HashMap<u64, LLVMMetadataRef>,
-    skipped_types: Vec<String>,
+    skipped_types: Vec<Cow<'a, str>>,
 }

 // Sanitize Rust type names to be valid C type names.
@@ -59,8 +59,8 @@ fn sanitize_type_name<T: AsRef<str>>(name: T) -> String {
     n
 }

-impl DISanitizer {
-    pub fn new(context: LLVMContextRef, module: LLVMModuleRef) -> DISanitizer {
+impl<'a> DISanitizer<'a> {
+    pub fn new(context: LLVMContextRef, module: LLVMModuleRef) -> DISanitizer<'a> {
         DISanitizer {
             context,
             module,
@@ -81,7 +81,7 @@ impl DISanitizer {
                     DW_TAG_structure_type => {
                         let (original_name, sanitized_name) = match di_composite_type.name() {
                             Some(name) => {
-                                let original_name = name.to_string_lossy().to_string();
+                                let original_name = name.to_string_lossy();
                                 let sanitized_name = sanitize_type_name(&original_name);

                                 (Some(original_name), Some(sanitized_name))
@@ -115,7 +115,7 @@ impl DISanitizer {

                                             let name = match original_name {
                                                 Some(name) => name,
-                                                None => "(anon)".to_owned(),
+                                                None => Cow::Borrowed("(anon)"),
                                             };
                                             let filename = match filename {
                                                 Some(filename) => {
```

One of the annoying errors that occur is:

```
error[E0502]: cannot borrow `di_composite_type` as mutable because it is also borrowed as immutable
   --> src/llvm/di.rs:176:29
    |
62  | impl<'a> DISanitizer<'a> {
    |      -- lifetime `'a` defined here
...
82  |                         let (original_name, sanitized_name) = match di_composite_type.name() {
    |                                                                     ----------------- immutable borrow occurs here
...
130 |                                             self.skipped_types.push(name);
    |                                             ----------------------------- argument requires that `di_composite_type` is borrowed for `'a`
...
176 |                             di_composite_type.replace_elements(sorted_elements);
    |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ mutable borrow occurs here
```

Well, the obvious solution would be pushing owned Strings to the vector to not cause an immutable borrow. Which pretty much brings us back to what I'm doing.

And we are not even benefiting from copy-on-write in this code anyway.
</blockquote></details>

Sure. I guess I was confused because the promotion to owned string happened in a different commit from where the need was introduced.
___
*[`src/llvm/di.rs` line 177 at r2](https://reviewable.io/reviews/aya-rs/bpf-linker/179#-NpoZCNx5dxpzxJBF_Xk:-NpouvSz2VRMAq0uTpBu:b-2bosum) ([raw file](https://github.com/aya-rs/bpf-linker/blob/44b1e400534cdf9eafbb64a9342cb991b37f7d8d/src/llvm/di.rs#L177)):*
<details><summary><i>Previously, vadorovsky (Michal Rostecki) wrote…</i></summary><blockquote>

It's explained in an another comment above (line 148). Do you think it's worth repeating here?
</blockquote></details>

I think it's better not to encode knowledge at a distance.

**Review by vadorovsky — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/aya-rs/bpf-linker/179)* status: 5 of 7 files reviewed, 6 unresolved discussions (waiting on @alessandrod and @tamird)

___
*[`src/bin/bpf-linker.rs` line 65 at r1](https://reviewable.io/reviews/aya-rs/bpf-linker/179#-NpoWlJt4hr5LLSYSuZU:-Nq4GYfQ4QP37-8fP5j8:ba0g51t) ([raw file](https://github.com/aya-rs/bpf-linker/blob/1f6476981267a1992d4a718b0c2a94b256ee7624/src/bin/bpf-linker.rs#L65)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Why can't you use it directly? Isn't `FromStr` all you need?
</blockquote></details>

OK, you are right, I'm dumb. Fixed.
___
*[`src/llvm/di.rs` line 80 at r2](https://reviewable.io/reviews/aya-rs/bpf-linker/179#-NpoYxqD9cpUXX53emhV:-Nq4GiDeACnsl3n0VCVZ:b-896fix) ([raw file](https://github.com/aya-rs/bpf-linker/blob/44b1e400534cdf9eafbb64a9342cb991b37f7d8d/src/llvm/di.rs#L80)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

You could write `if let Some((_, foo)) = ...` which seems fine
</blockquote></details>

Done.

File: src/bin/bpf-linker.rs (hunk 1 above)
Comment (vadorovsky) [reply src/bin/bpf-linker.rs]: Thanks for pointing out, it's inaccurate. I removed `off` from the comment.

File: src/bin/bpf-linker.rs (hunk 2 above)
Comment (vadorovsky) [reply src/bin/bpf-linker.rs:201]: Well, I decided to go with Tamir's suggestion and remove that enum completely. Now I'm just defining the argument as `Option<tracing::Level>`. If the argument is not provided (default case), logging is off. Do you agree with that solution?

**Review by tamird — APPROVED** [review]
> Reviewed 2 of 2 files at r9, 1 of 1 files at r10, 1 of 1 files at r11, 1 of 1 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/aya-rs/bpf-linker/179)* status: all files reviewed, 2 unresolved discussions (waiting on @alessandrod and @dave-tucker)

**Pull request merged**